src/rdkafka_cgrp.c

```diff
@@ -738,18 +949,297 @@ rd_kafka_rebalance_op (rd_kafka_cgrp_t *rkcg,
+
+        if (!a || !b)
+                return intersection;
+
+        RD_MAP_FOREACH(k, a_v, a) {
+                const PartitionMemberInfo_t *b_v = RD_MAP_GET(b, k);
+
+                if (b_v == NULL)
+                        continue;
+
+                RD_MAP_SET(intersection,
+                           rd_kafka_topic_partition_copy(k),
+                           PartitionMemberInfo_new(b_v->member,
+                                                   (rd_bool_t)
+                                                   (b_v->member ==
+                                                    a_v->member)));
+        }
+
+        return intersection;
+}
+
+
+/**
+ * @brief Set subtraction. Returns a set of all elemets of \p a
+ *        that are not elements of \p b. Sets the member field in
+ *        elements in the returned set to equal that of the
+ *        corresponding element in \p a
+ */
+map_toppar_member_info_t *
+rd_kafka_member_partitions_subtract (
+                map_toppar_member_info_t *a,
+                map_toppar_member_info_t *b) {
+        const rd_kafka_topic_partition_t *k;
+        const PartitionMemberInfo_t *a_v;
+        map_toppar_member_info_t *difference =
+                rd_calloc(1, sizeof(*difference));
+
+        RD_MAP_INIT(
+                difference,
+                RD_MAP_CNT(a),
+                rd_kafka_topic_partition_cmp,
+                rd_kafka_topic_partition_hash,
+                rd_kafka_topic_partition_destroy_free,
+                PartitionMemberInfo_free);
+
+        RD_MAP_FOREACH(k, a_v, a) {
+                const PartitionMemberInfo_t *b_v = RD_MAP_GET(b, k);
+
+                if (!b_v)
+                        RD_MAP_SET(difference,
+                                   rd_kafka_topic_partition_copy(k),
+                                   PartitionMemberInfo_new(a_v->member,
+                                                           rd_false));
+        }
+
+        return difference;
+}
+
+
+static void cooperative_protocol_adjust_assignment (
+                                        rd_kafka_cgrp_t *rkcg,
+                                        rd_kafka_group_member_t *members,
+                                        int member_cnt) {
+
+        /* https://cwiki.apache.org/confluence/display/KAFKA/KIP-429%3A+Kafk\
+           a+Consumer+Incremental+Rebalance+Protocol */
+
+        int i;
+        int total_assigned = 0;
+        int not_revoking = 0;
+        size_t par_cnt = 0;
+        const rd_kafka_topic_partition_t *toppar;
+        const PartitionMemberInfo_t *pmi;
+        map_toppar_member_info_t *assigned;
+        map_toppar_member_info_t *owned;
+        map_toppar_member_info_t *maybe_revoking;
+        map_toppar_member_info_t *ready_to_migrate;
+        map_toppar_member_info_t *unknown_but_owned;
+
+        for (i = 0 ; i<member_cnt ; i++)
+                par_cnt += members[i].rkgm_owned->cnt;
+
+        assigned = rd_kafka_collect_partitions(members,
```

Comment:
ok. what you say makes more structural sense, yes. ok. but nitpicker.

Reply:
no, your way is more code / ugly, mostly because you need to do the inner loop twice.

I do admit I don't like the bool much the way I have it (in a different language i might make that a delegate), but it's really easy to read, and doing the iteration twice over the members is a non-issue perf wise.